azure_costs_exporter/scrape.py

```diff
@@ -53,17 +53,27 @@ def convert_json_df(data):
 
     return df
 
-def extract_metrics_from_df(df, counter):
+def extract_metrics_from_df(df, metric_name, enrollment, registry):
     """
     Fill a counter object with the data from the data frame
     """
     groups = df.groupby(base_columns).sum()
-    for name, value in groups.iterrows():
-        meta = dict(zip(base_columns, name))
-        counter.record(value.ExtendedCost, **meta)
 
+    class AzureCollector(object):
+        def collect(self):
+            c = CounterMetricFamily(metric_name,
+                                    "Costs billed to Azure Enterprise Agreement {}".format(enrollment),
+                                    labels=base_columns)
 
-def query_metrics(enrollment, token, metric_name, month=None):
+            for name, value in groups.iterrows():
+                c.add_metric(name, value.ExtendedCost)
+
+            yield c
+
+    registry.register(AzureCollector())
```

Comment:
I would have loved to do the registration once or with the global REGISTRY object. However when testing this it turned out that the metrics output generated with generate_latest is tied to the lifetime of the registry.
When using a global object, the second call to metrics will contain all metrics twice. The third call three times the content etc.

Therefore it was on purpose to give the registry object a lifetime that is bound to the request.

@StephanErb have you tried that before and know where my mistake is?